### PR TITLE
add multivariate normal distribution with avf control variate

### PR DESCRIPTION
**Pull request by martinjankowiak**

### CODE REVIEW

File: pyro/distributions/avf_mvn.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    A distribution over vectors in which all the elements have a joint Gaussian density.
+
+    :param torch.Tensor loc: D-dimensional mean vector.
+    :param torch.Tensor scale_tril: Cholesky of Covariance matrix; D x D matrix.
+    :param torch.Tensor CV: 2 x L x D tensor that parameterizes the control variate; L is an arbitrary positive integer.
+    This parameter needs to be learned (i.e. adapted) to achieve lower variance gradients.
+
+    Example usage::
+
+    CV = torch.tensor(0.1 * torch.ones(2, 1, D), requires_grad=True)
```
Comment (jpchen) [inline pyro/distributions/avf_mvn.py]: nit: we should follow python and pyro convention and make these lower case with a more descriptive name eg `control_var`

File: pyro/distributions/avf_mvn.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    opt_cv = pyro.optim.Adam([CV], lr=0.1, betas=(0.5, 0.999))
+
+    for _ in range(1000):
+        d = AVFMultivariateNormal(loc, scale_tril, CV)
+        z = d.rsample()
+        cost = torch.pow(z, 2.0).sum()
+        cost.backward()
```
Comment (jpchen) [inline pyro/distributions/avf_mvn.py]: this looks runnable do you want to add it to doctests? via `>>>`

File: pyro/distributions/avf_mvn.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        BC_lab = B.unsqueeze(-1) * C.unsqueeze(-2)
+        diff_L_ab += (xi_ab.unsqueeze(0) * BC_lab).sum(0)
+        L_grad = torch.tril(diff_L_ab)
+
+        # compute CV grads
+        diff_B = (L_grad.unsqueeze(0) * C.unsqueeze(-2) * xi_ab.unsqueeze(0)).sum(2)
+        diff_C = (L_grad.t().unsqueeze(0) * B.unsqueeze(-2) * xi_ab.t().unsqueeze(0)).sum(2)
+        diff_CV = torch.stack([diff_B, diff_C])
+
+        return loc_grad, L_grad, diff_CV, None
```
Comment (jpchen) [inline pyro/distributions/avf_mvn.py]: what is the fourth value returned?

File: pyro/distributions/avf_mvn.py (hunk 1 above)
Comment (martinjankowiak) [reply pyro/distributions/avf_mvn.py]: ok

File: pyro/distributions/avf_mvn.py (hunk 2 above)
Comment (martinjankowiak) [reply pyro/distributions/avf_mvn.py]: nope purposely didn't do that

File: pyro/distributions/avf_mvn.py (hunk 3 above)
Comment (martinjankowiak) [reply pyro/distributions/avf_mvn.py]: the shape param has no grad

File: pyro/distributions/avf_mvn.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    :param torch.Tensor loc: D-dimensional mean vector.
+    :param torch.Tensor scale_tril: Cholesky of Covariance matrix; D x D matrix.
+    :param torch.Tensor control_var: 2 x L x D tensor that parameterizes the control variate;
+    L is an arbitrary positive integer.
+    This parameter needs to be learned (i.e. adapted) to achieve lower variance gradients.
+
+    Example usage::
+
+    control_var = torch.tensor(0.1 * torch.ones(2, 1, D), requires_grad=True)
+    opt_cv = pyro.optim.Adam([control_var], lr=0.1, betas=(0.5, 0.999))
```
Comment (fritzo) [inline pyro/distributions/avf_mvn.py]: Do you mean `torch.optim.Adam` here? Doctests may have caught this error, just sayin.

File: pyro/distributions/avf_mvn.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        cost.backward()
+        opt_cv.step()
+        opt_cv.zero_grad()
+
+    """
+    arg_constraints = {"loc": constraints.real, "scale_tril": constraints.lower_triangular,
+                       "control_var": constraints.real}
+
+    def __init__(self, loc, scale_tril, control_var):
+        assert(loc.dim() == 1), "AVFMultivariateNormal loc must be 1-dimensional"
```
Comment (fritzo) [inline pyro/distributions/avf_mvn.py]: Use `if ...: raise ValueError(...)` for user facing error messages. `assert` should be for developer-facing errors.

File: pyro/distributions/avf_mvn.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        assert(scale_tril.dim() == 2), "AVFMultivariateNormal scale_tril must be 2-dimensional"
+        assert control_var.dim() == 3, \
+            "control_var should be of size 2 x L x D, where D is the dimension of the location parameter loc"
+        assert control_var.size(0) == 2, \
+            "control_var should be of size 2 x L x D, where D is the dimension of the location parameter loc"
+        assert control_var.size(2) == loc.size(0), \
+            "control_var should be of size 2 x L x D, where D is the dimension of the location parameter loc"
+        self.control_var = control_var
+        super(AVFMultivariateNormal, self).__init__(loc, scale_tril=scale_tril)
+        self.loc = loc
```
Comment (fritzo) [inline pyro/distributions/avf_mvn.py]: Why are you setting `.loc` and `.scale` after `super(...).__init__()` has already set them?

File: tests/distributions/test_omt_mvn.py [hunk 7]
```diff
@@ -18,31 +17,81 @@ def analytic_grad(L11=1.0, L22=1.0, L21=1.0, omega1=1.0, omega2=1.0):
...
+    computed_grads = []
+
+    for _ in range(n_samples):
+        z = dist.rsample()
+        torch.cos((omega*z).sum(-1)).mean().backward()
+        assert(off_diag.grad.size() == off_diag.size())
+        assert(loc.grad.size() == loc.size())
+        assert(torch.triu(off_diag.grad, 1).sum() == 0.0)
+
+        computed_grad = off_diag.grad.clone().data.numpy()[1, 0]
```
Comment (fritzo) [inline tests/distributions/test_omt_mvn.py]: Replace `.clone()` with `.cpu()`?

File: tests/distributions/test_omt_mvn.py [hunk 8]
```diff
@@ -18,31 +17,81 @@ def analytic_grad(L11=1.0, L22=1.0, L21=1.0, omega1=1.0, omega2=1.0):
...
+        dist = OMTMultivariateNormal(loc, L)
+    elif mvn_dist == 'AVFMultivariateNormal':
+        CV = torch.tensor(1.1 * torch.rand(2, k, 3), requires_grad=True)
+        dist = AVFMultivariateNormal(loc, L, CV)
+
     z = dist.rsample(sample_shape)
     torch.cos((omega*z).sum(-1)).mean().backward()
 
-    computed_grad = off_diag.grad.data.cpu().numpy()[1, 0]
+    computed_grad = off_diag.grad.clone().data.numpy()[1, 0]
```
Comment (fritzo) [inline tests/distributions/test_omt_mvn.py]: The `.cpu()` is needed for cuda tests. What is your reason for removing it?

File: pyro/distributions/avf_mvn.py (hunk 5 above)
Comment (martinjankowiak) [reply pyro/distributions/avf_mvn.py]: done

File: pyro/distributions/avf_mvn.py (hunk 6 above)
Comment (martinjankowiak) [reply pyro/distributions/avf_mvn.py]: fixed

File: tests/distributions/test_omt_mvn.py (hunk 8 above)
Comment (martinjankowiak) [reply tests/distributions/test_omt_mvn.py]: i need the clone so that it doesn't get zeroed out

Comment (fritzo) [reply tests/distributions/test_omt_mvn.py]: I think you could instead `off_diag.grad.cpu()[1, 0].item()`, and that would work in cuda as well

**Pull request merged**